fec-bareminimum/project-catwalk
# test provided values from Context Provider

These tests are a hacky way to test the values being passed down to children from the Context Provider.

### When using the Context:

1. The Child should be able to destructure this value from ```useProducts```. 
   - test that the value is not undefined
   - ```useProducts``` should export this key in the values object

2. The Child will expect a specific data type to work with
   -  test the INSTANCE of the value (array can be type "object")
   - ```ProductsContext``` should store the proper value type in state

### How to test:
 - Create a mock component to use in the ```describe``` block.
 - In the MOCK child: Access the Context with the ```useProducts()``` hook
 - Write assertions in the Child Component code block
 ```js
  const MockChild = () => {
     const productExports = useProducts()

     expect(productExports["productList"]).not.toBe(undefined)
     expect(productExports["productList"] instanceof Array).toBeTruthy()

     return <p>test</p>
  }
```
- NOW: ```render``` the MockChild
  - the ProductContext values are inaccessible **outside** of this component

### Example:
```js
describe("ProductsContext", () => {
  test("passes correct values to children", () => {
    const MockChild = () => {
      const productExports = useProducts() // return object from useProducts function

      // iterate over the expected exports to assert the key matches  the desired instance
      expectedExports.forEach(({ key, targetInstance }) => {
        try {
          expect(productExports[key]).not.toBe(undefined)
        } catch {
          console.error(`expected ${key} to NOT be undefined`)
        }

        try {
          expect(productExports[key] instanceof targetInstance).toBeTruthy()
        } catch {
          console.error(`expected ${key} to be instance of ${targetInstance}`)
        }
      })

      return <p>test</p>
    }

    render(<MockChild />)
  })
})

const expectedExports = [
  {
    key: "fetchCartProducts",
    targetInstance: Function,
  },
  ...
]
```

## NOTE:
Jest may say the test passes. Without a ```try catch``` block, the error stack message will clutter up the terminal.
By running the ```console.error("custom assertion message")```, it is easier to see why this type of test is failing.
